presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileWriter.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.pagefile;
+
+import com.facebook.presto.hive.HiveFileWriter;
+import com.facebook.presto.orc.OrcDataSink;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.page.PagesSerde;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.concurrent.Callable;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_CLOSE_ERROR;
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_DATA_ERROR;
+import static java.util.Objects.requireNonNull;
+
+public class PageFileWriter
+        implements HiveFileWriter
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(PageFileWriter.class).instanceSize();
+
+    private final PageWriter pageWriter;
+    private final Callable<Void> rollbackAction;
+    private final PagesSerde pagesSerde;
+
+    public PageFileWriter(
+            OrcDataSink dataSink,
+            PagesSerde pagesSerde,
+            Callable<Void> rollbackAction)
+    {
+        pageWriter = new PageWriter(dataSink);
+        this.pagesSerde = requireNonNull(pagesSerde, "pagesSerde is null");
+        this.rollbackAction = requireNonNull(rollbackAction, "rollbackAction is null");
+    }
+
+    @Override
+    public long getWrittenBytes()
+    {
+        return pageWriter.getWrittenBytes();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return INSTANCE_SIZE + pageWriter.getRetainedBytes();
+    }
+
+    @Override
+    public void appendRows(Page dataPage)
+    {
+        try {
+            pageWriter.write(pagesSerde.serialize(dataPage));
+        }
+        catch (IOException | UncheckedIOException e) {
+            throw new PrestoException(HIVE_WRITER_DATA_ERROR, e);
+        }
+    }
+
+    @Override
+    public void commit()
+    {
+        try {
+            pageWriter.close();
+        }
+        catch (IOException | UncheckedIOException e) {
+            try {
+                rollbackAction.call();
+            }
+            catch (Exception ignored) {
+                // ignore
+            }
+            throw new PrestoException(HIVE_WRITER_CLOSE_ERROR, "Error committing write to Hive", e);
+        }
+    }
+
+    @Override
+    public void rollback()
+    {
```

Comment:
Should the rollback action be run here as well?

Reply:
It looks like the `OrcFileWriter` does both, it closes the output file by closing the `pageWriter` and also runs the rollback action: https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriter.java#L176